ramips: add support for I-O DATA WNPR2600G

musashino205 opened the pull request:
I-O DATA WNPR2600G is a 2.4/5 GHz band 11ac router, based on MediaTek
MT7621A.

Specification:

- SoC		: MediaTek MT7621A
- RAM		: DDR3 128 MiB
- Flash		: SPI-NOR 16 MiB
- WLAN		: 2.4/5 GHz 4T4R
  - 2.4 GHz	: MediaTek MT7615
  - 5 GHz	: MediaTek MT7615
- Ethernet	: 10/100/1000 Mbps
  - Switch	: MT7530 (SoC)
- LED/keys	: 4x/3x (2x buttons, 1x slide-switch)
- UART		: through-hole on PCB
  - J1: Vcc, RX, GND, TX from SoC side
  - 57600n8

Flash instruction using factory image:

1. Boot WNPR2600G normaly
2. Access to "http://192.168.0.1/" and open firmware update page
("ファームウェア")
3. Select the OpenWrt factory image and click update ("更新") button
to perform firmware update
4. Wait ~150 seconds to complete flashing

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

981213 commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -220,6 +220,10 @@ iodata,wn-ac733gr3)
 	;;
+iodata,wnpr2600g)
+	ucidef_set_led_wlan "wlan2g" "WLAN2G" "$boardname:green:wlan2g" "phy0radio"
+	ucidef_set_led_wlan "wlan5g" "WLAN5G" "$boardname:green:wlan5g" "phy1radio"
```
These triggers can be set in dts using linux,default-trigger property.
BTW Is there a phy0tpt/phy1tpt trigger available? If there is, you could use tpt trigger which will also indicate active traffic.

musashino205 replied:
I checked *trigger* in each LEDs (/sys/class/leds/*/trigger), but there is no ```phy*tpt``` trigger.

>These triggers can be set in dts using linux,default-trigger property.

OK, I'll try it.

981213 commented on target/linux/ramips/dts/WNPR2600G.dts:
```diff
@@ -0,0 +1,168 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
10MHz is a pretty slow SPI frequency. You could try if a higher frequency works for you.
MTK uses 30MHz in their SDK. All my mt7621/mt7628 boards works fine with 40MHz.

musashino205 replied:
Sounds great, I'll try to use higher clock.

musashino205 replied:
Done, 40MHz works and reading from flash has become faster. (writing has not been tested)

```
# 10MHz
root@OpenWrt:/tmp/mtd# time dd if=/dev/mtdblock3 of=mtd3_firmware.bin
27904+0 records in
27904+0 records out
real    0m 12.92s
user    0m 0.03s
sys     0m 0.38s

# 40MHz
root@OpenWrt:/tmp/mtd# time dd if=/dev/mtdblock3 of=mtd3_firmware.bin
27904+0 records in
27904+0 records out
real    0m 4.40s
user    0m 0.01s
sys     0m 0.56s
```

981213 reviewed (commented):
Hi! two comments below:

adrianschmutzler commented on target/linux/ramips/dts/WNPR2600G.dts:
```diff
@@ -0,0 +1,168 @@
+			};
+
+			Factory: partition@40000 {
+				label = "Factory";
```
Never seen an upper-case label before. But DTS says it's allowed.
Consider converting (both) to lower-case like all the others.

musashino205 replied:
```Factory``` is used in several boards.
```
$ grep "label = \"Factory\"" target/linux/ramips/dts/*
target/linux/ramips/dts/M2M.dts:                                label = "Factory";
target/linux/ramips/dts/MIR3G.dts:                      label = "Factory";
target/linux/ramips/dts/MIR3P.dts:                      label = "Factory";
target/linux/ramips/dts/WN-AC1167GR.dts:                                label = "Factory";
target/linux/ramips/dts/WN-AC733GR3.dts:                                label = "Factory";
target/linux/ramips/dts/WN-AX1167GR.dts:                                label = "Factory";
target/linux/ramips/dts/WN-GX300GR.dts:                         label = "Factory";
target/linux/ramips/dts/WNPR2600G.dts:                          label = "Factory";
target/linux/ramips/dts/WR1201.dts:                             label = "Factory";
```

OK, I'll convert to lower-case.

adrianschmutzler replied:
Yes, I've seen it too, but also found lower-case examples. "The others" referred to the other labels in the same DTS in this case.

That's why I explicitly added that DTS allows it: It's just my personal belief that the upper-case is odd, but you are free to choose.

musashino205 replied:
I see. That partition is named as ```Factory``` in the OEM firmware, so I've used this name.

OEM log:
```
Creating 8 MTD partitions on "raspi":
0x000000000000-0x000001000000 : "ALL"
0x000000000000-0x000000030000 : "Bootloader"
0x000000030000-0x000000040000 : "Config"
0x000000040000-0x000000050000 : "Factory"
0x000000050000-0x000000df0000 : "Kernel"
0x000000220000-0x000000df0000 : "user"
0x000000df0000-0x000000f80000 : "manufacture"
0x000000f80000-0x000001000000 : "storage"
```

musashino205 commented:
UPDATE:
- use dt bindings for wireless-related LEDs instead of 01_leds
- use 40MHz for SPI-NOR flash
- use a lower-case label for factory partition

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -228,6 +228,18 @@ define Device/iodata_wn-gx300gr
 TARGET_DEVICES += iodata_wn-gx300gr
 
+define Device/iodata_wnpr2600g
+  DTS := WNPR2600G
```
Already suggested this in another PR:
I wonder whether it wouldn't be desirable to start with improving the naming of DTSes, e.g.
iodata_wnpr2600g.dts
mt7621_iodata_wnpr2600g.dts

However, this is just a suggestion without any obligation (don't even know whether devs will like it).

981213 replied:
I think we could probably change the default image template like ath79 before applying this naming scheme.

musashino205 replied:
I also like the style in ath79 target.
It is useful because it is easy to see which SoC the device has and is grouped by it in the list.

However, (if we do,) it may be better to do this collectively within another patchset rather than changing it with one model.

adrianschmutzler replied:
@981213 Yes, that would be obviously even better. Who will do it?
@musashino205 Okay, no objection if you keep the current state ...

981213 replied:
@adrianschmutzler I could do that the week after next if you don't beat me to it :)

adrianschmutzler replied:
Well, I just started and will wait for some feedback from devs now:
https://github.com/openwrt/openwrt/pull/2171

adrianschmutzler commented:
There has been a reorganization in ramips target today. Please:
- update your board.d subfiles since they have been re-sorted
- update your Makefile (mt****/rt****.mk) changes so sorting of devices is correct
- rename your DTS/DTSI files following the scheme soc_vendor_device.dts(i), e.g. mt7620a_tplink_archer-c50-v1.dts (and update includes there if necessary)
- remove `DTS := ...` from Makefile
- add `MTK_SOC` to Makefile (needs to match "soc" in DTS file name), e.g. `MTK_SOC := mt7620a`

Thank you!

musashino205 commented:
OK, I'll work for all my ramips-related PRs.

musashino205 committed:
ramips: add support for I-O DATA WNPR2600G

I-O DATA WNPR2600G is a 2.4/5 GHz band 11ac router, based on MediaTek
MT7621A.

Specification:

- SoC           : MediaTek MT7621A
- RAM           : DDR3 128 MiB
- Flash         : SPI-NOR 16 MiB
- WLAN          : 2.4/5 GHz 4T4R
  - 2.4 GHz     : MediaTek MT7615
  - 5 GHz       : MediaTek MT7615
- Ethernet      : 10/100/1000 Mbps
  - Switch      : MT7530 (SoC)
- LED/keys      : 4x/3x (2x buttons, 1x slide-switch)
- UART          : through-hole on PCB
  - J1: Vcc, RX, GND, TX from SoC side
  - 57600n8

Flash instruction using factory image:

1. Boot WNPR2600G normaly
2. Access to "http://192.168.0.1/" and open firmware update page
("ファームウェア")
3. Select the OpenWrt factory image and click update ("更新") button
to perform firmware update
4. Wait ~150 seconds to complete flashing

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 commented:
UPDATE:
- rebase from master

adrianschmutzler commented:
@blocktrron Maybe you want to deal with this PR, as it's lying around for some time already and seems to be in good shape.

blocktrron commented on target/linux/ramips/dts/mt7621_iodata_wnpr2600g.dts:
```diff
@@ -0,0 +1,165 @@
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
+	mediatek,portmap = "wllll";
```
I've had a look at the whole "mediatek,portmap" story some time ago. If I'm not mistaken, this will not work.

If you look at https://github.com/openwrt/openwrt/blob/34e22653ac18b6ac7fd368ca47625f665808067f/target/linux/ramips/files-4.14/drivers/net/ethernet/mediatek/esw_rt3050.c#L1366 , you will see that only 32 bit integers are valid here. It just "fails" silently.

So TL;DR: 
Use <0x3e> to fit the current scheme.

This value is however only "needed" for displaying a "AN flag in swconfig. It affects no functionality / configuration. Also - the mapping is a bit misleading, as currently a lot of devices have this property either declared incorrect or even incorrect (as a String).

To add insult to injury, the LAN flag is very often (so in my suggestion) set for the CPU port, which is strange as this port carries LAN and WAN separated by a VLAN tag.

So we should maybe consider dropping this property globally altogether, as we have a cleaner track in the userspace (and the property isn't used upstream).

musashino205 replied:
Thank you for your review. However, I recognize that ```of_property_read_string()``` is used in MT7530 switch, but is it wrong?

https://github.com/openwrt/openwrt/blob/920abb9b5865fe0994a5c8115cde616dac63d5ae/target/linux/ramips/files-4.14/drivers/net/ethernet/mediatek/mt7530.c#L238

blocktrron replied:
Okay, i was a bit blind there :) You are right, for the mt7620 it's a string then. Hooray for consistency.

However, if I'm not mistaken, the actual config is overwritten afterwards by userspace. In addition, only the predefined portmaps seem to be supported (But I'm not sure how much of an issue this is).

blocktrron commented on target/linux/ramips/image/Makefile:
```diff
@@ -90,6 +90,25 @@ define Build/relocate-kernel
+  ) > $(KDIR)/tmp/$(DEVICE_NAME).header
+  $(call Build/xor-image,-p $(xor_pattern) -x)
+  cat $(KDIR)/tmp/$(DEVICE_NAME).header $@ > $@.new
+  mv $@.new $@
```
This was missing before you moved it, but a `rm -rf $(KDIR)/tmp/$(DEVICE_NAME).header` would be a bit more clean here.

musashino205 replied:
Indeed, I'll add it.

adrianschmutzler replied:
I would like to have that as a separate commit, just before the device support (especially since we are changing stuff here now). Should be easy to do, though.

musashino205 commented:
UPDATE:
- add cleaning up ```header``` file in elx-header
- unify MAC address configuration with WN-AC1167GR and WN-AC733GR3

adrianschmutzler commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -607,7 +608,8 @@ ramips_setup_macs()
-	iodata,wn-ac733gr3)
+	iodata,wn-ac733gr3|\
+	iodata,wnpr2600g)
 		wan_mac=$(macaddr_add "$(mtd_get_mac_binary factory 4)" -1)
```
Is this somewhere on flash, too, or is it really necessary to calculate?

musashino205 replied:
Replaced to ```wanaddr``` in u-boot-env.

musashino205 committed:
ramips: replace MAC address configuration for I-O DATA WN-AC-GR devices

following I-O DATA devices have a MAC address as "wanaddr" for WAN
interface in u-boot-env:

- WN-AC1167GR
- WN-AC733GR3

I replaced MAC address configuration in these devices to omit
address calculation in 02_network.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 commented:
UPDATE:

- use ```mtd_get_mac_ascii``` instead of ```mtd_get_mac_binary```
- rebase from master

adrianschmutzler commented on target/linux/ramips/dts/mt7621_iodata_wnpr2600g.dts:
```diff
@@ -0,0 +1,165 @@
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
```
Did you check whether there is something in the typical locations? 0x28 0x2e 0xe000 0xe006 ?

musashino205 replied:
nothing

adrianschmutzler replied:
Okay, fine then.

adrianschmutzler commented:
Thanks for the MAC address cleanup.

I'd still prefer a separate commit for moving elx-header...

musashino205 committed:
ramips: move "elx-header" to Makefile from mt7620.mk and mt7621.mk

I-O DATA WNPR2600G has an "elx-header", so move this definition to
generic makefile to use it from mt7621 subtarget.
This definition is also added to mt7621.mk in
f285e8634c57d28aa970b80c5c59e85485f35c7d, so remove it from mt7621.mk.

And added a line to cleanup used header file.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 commented:
UPDATE:

- split the commit (elx-header)

adrianschmutzler commented:
@musashino205 Please add a short justification for the added line to the elx-header to the commit message
@blocktrron This should be ready for another/final review.

musashino205 commented:
done

musashino205 commented:
UPDATE:
- rebase from master to resolve conflicts
- also remove ```Build/elx-header``` from mt7621.mk
  - added in f285e8634c57d28aa970b80c5c59e85485f35c7d

adrianschmutzler commented:
LGTM. Which MAC address is on the label?

musashino205 commented:
WAN's address

adrianschmutzler commented:
> WAN's address

Thanks. Then please add it to 02_network (you have to create a new case then):
```
	iodata,wnpr2600g)
		wan_mac=$(mtd_get_mac_ascii u-boot-env wanaddr)
		label_mac=wan_mac
		;;
```

musashino205 commented:
done

981213 commented:
Merged into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git
Thanks!
